### PR TITLE
DROOLS-2349: [DMN Editor] Decision Navigator dock

### PR DESCRIPTION
**Pull request by karreiro**
See https://issues.jboss.org/browse/DROOLS-2349

---

- Expand/Collapse demo:
expand-collapse

- Navigation demo:
navigation-nested-and-top-level-nodes

- Nodes alphabetically ordered demo:
order-do-redo

---

The [DecisionNavigatorPresenter](https://github.com/karreiro/kie-wb-common/blob/5976f65c03e16ba420e74ba5e3fe7d2117420d72/kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/DecisionNavigatorPresenter.java), the [DecisionNavigatorDock](https://github.com/karreiro/kie-wb-common/blob/5976f65c03e16ba420e74ba5e3fe7d2117420d72/kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/DecisionNavigatorDock.java), and [DecisionNavigatorObserver](https://github.com/karreiro/kie-wb-common/blob/5976f65c03e16ba420e74ba5e3fe7d2117420d72/kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/DecisionNavigatorObserver.java) are a good start to the review.

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorTest.java [hunk 1]
```diff
@@ -103,7 +112,21 @@ public void testExitWithCommand() {
...
         testedEditor.exit();
 
+        verify(decisionNavigator).clearSelections();
         verify(editorToolbar, atLeast(1)).enable(any(ToolbarCommand.class));
         verify(command).execute();
+        assertEquals(Optional.empty(), testedEditor.getExitCommand());
+    }
+
+    @Test
+    public void testOnCanvasFocusedSelectionEvent() {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorTest.java]: I'd be interested to learn of the situation when the User can select a canvas item and the "grid" editor is displayed.

IIUC this code (well, the code being tested here!) exits the "grid" editor when a Canvas item gets the focus?

**Comment by manstis** [comment]
Sorry @karreiro but I broke your PR :-(

**Comment by karreiro** [comment]
No problem @manstis 😊

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorTest.java (hunk 1 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorTest.java]: Precisely, this code exits the grid editor when a canvas node gets the focus.

When a node is clicked, we're focusing on the graph node, and if grid editor is opened, it's closed to give place to the canvas graph :-)

See:
demo

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorTest.java]: Ah.. @karreiro now I understand. Thanks :-)

**Comment by jomarko** [comment]
manual check finished, lower priority issues linked in the original jira

**Comment by manstis** [comment]
@karreiro Hi, I had a quick play and there seem to be issues relating to "undo" - I added a ```Decision```, edited it, changing it to a ```LiteralExpression``` and the Navigator showed both the ```Decision``` and ```LiteralExpression```. If I "undo" the setting of the ```LiteralExpression``` the Navigator still showed it as a child of the ```Decision```. If I further "undo" navigation to the ```Decision``` grid the Navigator continues to show both the ```Decision``` and ```LiteralExpression```. It is only when I "undo" creation of the ```Decsion``` that both the ```Decision``` and ```LiteralExpression``` disappear.

Did you want me to review the code changes you've made so far; or did you want to look at this and the issues raised by @jomarko before? IDK what impact the issues may have on the code you've written and I would not want to review this PR further if you expect there to be a lot of changes.

Please advise.. shall I review, or wait? Thanks.

**Comment by karreiro** [comment]
@manstis Thank you for the quick play. I don't think that I'll need to make huge changes to fix the issues that Jozef raised.

However, this is a big PR, so I prefer to be conservative. **Wait for my "ok" to review this**.

Thanks.

**Comment by karreiro** [comment]
Thank you for the feedback @manstis and @jomarko.

I just fixed the bugs that you reported, see:
2018-04-04 10 47 05 pm
2018-04-04 10 41 52 pm

So, could please take a look? :-)

Regarding the reported issues ([DROOLS-2433](https://issues.jboss.org/browse/DROOLS-2433), [DROOLS-2432](https://issues.jboss.org/browse/DROOLS-2432), [DROOLS-2434](https://issues.jboss.org/browse/DROOLS-2434), [DROOLS-2435](https://issues.jboss.org/browse/DROOLS-2435), [DROOLS-2431](https://issues.jboss.org/browse/DROOLS-2431)), I think that I can take care of them in other PRs. WDYT?

Thank you.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
(I just rebased the branch) ;-)

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
I just rebased the branch.

**Comment by manstis** [comment]
@karreiro Do we need ```kie-wb-common-ala/kie-wb-common-ala-distribution/audit.log```?

I could not add a review comment against the file.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java [hunk 2]
```diff
@@ -322,8 +326,20 @@ public Layer getLayer() {
         return layer;
     }
 
+    @Override
+    public void select() {
+        fireExpressionEditorChanged();
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java]: A possible problem (not that I experienced it when running - so possibly nothing to worry about) is that ```DMNGridLayer.select(GridWidget)``` calls ```BaseGridWidget.select()``` (and ```BaseGridWidget.deselect()``` see L#341 below) so the ```ExpressionEditorChanged``` event will trigger X times when a grid is selected (where X is the number of grids on the layer).. So just not optimal.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java [hunk 3]
```diff
@@ -322,8 +326,20 @@ public Layer getLayer() {
...
+        super.select();
+    }
+
+    private void fireExpressionEditorChanged() {
+        editorSelectedEvent.fire(new ExpressionEditorChanged());
+    }
+
     @Override
     public void deselect() {
+        fireExpressionEditorChanged();
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java]: See my comment above.

**Comment by manstis** [comment]
@karreiro One other comment is that isn't it normally customary to select the expand/collapse icon to expand/collapse a tree node with children as opposed to the tree node's icon? See http://www.patternfly.org/pattern-library/widgets/#bootstrap-tree-view

**Comment by karreiro** [comment]
@manstis no, we don't need it. It's a generated file, which I accidentally pushed (sorry).

It's fixed now :-)

**Comment by manstis** [comment]
@karreiro All the items I raise can be the subject of new JIRAs so we can get this merged.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java (hunk 2 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java]: @manstis Yes, definitely it's not optimal. The `DecisionNavigatorObserver` logic is not affected by this, but we certainly should improve this (could we do it in another PR?).

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/BaseExpressionGrid.java]: @karreiro Yes, sure, my last comment was that any improvements could be the subject of new JIRA(s).

**Comment by karreiro** [comment]
@manstis thank you. I just raised the JIRAs related to the issues that you found: [DROOLS-2464](https://issues.jboss.org/browse/DROOLS-2464), [DROOLS-2465](https://issues.jboss.org/browse/DROOLS-2465) ;-)

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
@manstis 🎉we're green! :-)

**Pull request merged**